cql3/statements/ks_prop_defs.cc

```diff
@@ -74,20 +75,64 @@ std::optional<sstring> ks_prop_defs::get_replication_strategy_class() const {
     return _strategy_class;
 }
 
-lw_shared_ptr<keyspace_metadata> ks_prop_defs::as_ks_metadata(sstring ks_name) {
+std::map<sstring, sstring> ks_prop_defs::prepare_options(const sstring& strategy_class, const std::map<sstring, sstring>& old_options = {}) {
     auto options = get_replication_options();
     options.erase(REPLICATION_STRATEGY_CLASS_KEY);
-    return keyspace_metadata::new_keyspace(ks_name, get_replication_strategy_class().value(), options, get_boolean(KW_DURABLE_WRITES, true));
+
+    if (strategy_class != "NetworkTopologyStrategy") {
+        return options;
+    }
+
+    // For users' convenience, expand the 'replication_factor' option into a replication factor for each DC.
+    // If the user simply switches from another strategy without providing any options,
+    // but the other strategy used the 'replication_factor' option, it will also be expanded.
+    // See issue CASSANDRA-14303.
+
+    sstring rf;
+    auto it = options.find(REPLICATION_FACTOR_KEY);
+    if (it != options.end()) {
+        rf = it->second;
+        options.erase(it);
+    } else if (options.empty()) {
+        auto it = old_options.find(REPLICATION_FACTOR_KEY);
+        if (it != old_options.end()) {
+            rf = it->second;
+        }
+    }
+
+    if (!rf.empty()) {
+        // We keep previously specified DC factors for safety.
+        for (const auto& opt: old_options) {
+            if (opt.first != REPLICATION_FACTOR_KEY) {
+                options.insert(opt);
+            }
+        }
+
+        for (const auto& dc: service::get_local_storage_service().get_token_metadata().
```

Comment:
:(

Reply:
Would `const auto& datacenters = service::stuff()` before the loop and `for (const auto& dc: datacenters)` change the ":(" into a ":)"?

Reply:
The :( refers to accessing a global object (get_local_storage_service()). Is there away to get the token_metadata without a global reference? We declared a war on globals some time ago as they hide dependencies and make unit testing difficult. If not, we'll keep the :( and pass it to whoever removes that global eventually.

btw, our coding style puts spaces before and after ":" in range for.

Reply:
>  Is there a way to get the token_metadata without a global reference?

There is, but not without some "hard" work. Here's what I'd do:

1. provide a reference to _the_storage_service in qp.start in main.cc
2. store service::storage_service& _service in query_processor, similarly to service::storage_proxy& _proxy
3. pass _service to cql_statement::execute in query_processor::process and friends
4. pass the service further down to schema_altering_statement::announce_migration in execute

Should I open an issue? I don't think those refactors belong in this pull request.
For now, I moved the get_local_storage_service() call higher up in the call stack, to announce_migration. This is where other globals are accessed and should make future refactoring easier.

By the way: a reference to storage_proxy is already being stored in query_processor and passed down to execute, announce_migration, and validate in schema_altering_statement(s). But sometimes it gets ignored and get_local_storage_proxy() gets called nevertheless... (see, for example, alter_keyspace_statement::announce_migration). I didn't change it in this PR so to not mix things up.